agent/proxycfg/testing.go

```diff
@@ -0,0 +1,246 @@
+// leaf cache resonse, signed by an existing CA.
+func TestLeafForCA(t testing.T, ca *structs.CARoot) *structs.IssuedCert {
+	leafPEM, pkPEM := connect.TestLeaf(t, "web", ca)
+
+	leafCert, err := connect.ParseCert(leafPEM)
+	require.NoError(t, err)
+
+	return &structs.IssuedCert{
+		SerialNumber:  connect.HexString(leafCert.SerialNumber.Bytes()),
+		CertPEM:       leafPEM,
+		PrivateKeyPEM: pkPEM,
+		Service:       "web",
+		ServiceURI:    leafCert.URIs[0].String(),
+		ValidAfter:    leafCert.NotBefore,
+		ValidBefore:   leafCert.NotAfter,
+	}
+}
+
+// TestIntentions returns a sample intentions match result useful to
+// mocking service discovery cache results.
+func TestIntentions(t testing.T) *structs.IndexedIntentionMatches {
+	return &structs.IndexedIntentionMatches{
+		Matches: []structs.Intentions{
+			[]*structs.Intention{
+				&structs.Intention{
+					ID:              "foo",
+					SourceNS:        "default",
+					SourceName:      "billing",
+					DestinationNS:   "default",
+					DestinationName: "web",
+					Action:          structs.IntentionActionAllow,
+				},
+			},
+		},
+	}
+}
+
+// TestUpstreamNodes returns a sample service discovery result useful to
+// mocking service discovery cache results.
+func TestUpstreamNodes(t testing.T) structs.CheckServiceNodes {
+	return structs.CheckServiceNodes{
+		structs.CheckServiceNode{
+			Node: &structs.Node{
+				ID:         "test1",
+				Node:       "test1",
+				Address:    "10.10.1.1",
+				Datacenter: "dc1",
+			},
+			Service: structs.TestNodeService(t),
+		},
+		structs.CheckServiceNode{
+			Node: &structs.Node{
+				ID:         "test2",
+				Node:       "test2",
+				Address:    "10.10.1.2",
+				Datacenter: "dc1",
+			},
+			Service: structs.TestNodeService(t),
+		},
+	}
+}
+
+// TestConfigSnapshot returns a fully populated snapshot
+func TestConfigSnapshot(t testing.T) *ConfigSnapshot {
+	roots, leaf := TestCerts(t)
+	return &ConfigSnapshot{
+		ProxyID: "web-sidecar-proxy",
+		Address: "0.0.0.0",
+		Port:    9999,
+		Proxy: structs.ConnectProxyConfig{
+			DestinationServiceID:   "web",
+			DestinationServiceName: "web",
+			LocalServiceAddress:    "127.0.0.1",
+			LocalServicePort:       8080,
+			Config: map[string]interface{}{
+				"foo": "bar",
+			},
+			Upstreams: structs.TestUpstreams(t),
+		},
+		Roots: roots,
+		Leaf:  leaf,
+		UpstreamEndpoints: map[string]structs.CheckServiceNodes{
+			"service:db": TestUpstreamNodes(t),
+		},
+	}
+}
+
+// ControllableCacheType is a cache.Type that simulates a typical blocking RPC
+// but lets us controll the responses and when they are deliverd easily.
+type ControllableCacheType struct {
+	index uint64
+	value atomic.Value
+	// Need a condvar to trigger all blocking requests (there might be multiple
+	// for same type due to background refresh and timing issues) when values
+	// change. Chans make it nondeterministic which one triggers or need extra
+	// locking to coodrinate rplacing after close etc.
```

Comment:
`s/coordrinate rplacing/coordinate replacing/`

Reply:
This is how all the cool kids are spelling it these days 😉